h2o-core/src/main/java/hex/ModelBuilder.java

```diff
@@ -1227,14 +1229,50 @@ protected Frame rebalance(final Frame original_fr, boolean local, final String n
     return rebalanced_fr;
   }
 
+  private double rebalanceRatio() {
+    String mode = H2O.getCloudSize() == 1 ? "single" : "multi";
+    String ratioStr = System.getProperty(H2O.OptArgs.SYSTEM_PROP_PREFIX + "rebalance.ratio." + mode, "1.0");
+    return Double.parseDouble(ratioStr);
+  }
+
   /**
    * Find desired number of chunks. If fewer, dataset will be rebalanced.
    * @return Lower bound on number of chunks after rebalancing.
    */
   protected int desiredChunks(final Frame original_fr, boolean local) {
-    return Math.min((int) Math.ceil(original_fr.numRows() / 1e3), H2O.NUMCPUS);
+    if (H2O.getCloudSize() > 1 && Boolean.getBoolean(H2O.OptArgs.SYSTEM_PROP_PREFIX + "rebalance.enableMulti"))
+      return desiredChunk_multi(original_fr);
+    else
+      return desiredChunk_single(original_fr);
+  }
+
+  // single-node version (original version)
+  private int desiredChunk_single(final Frame originalFr) {
+    return Math.min((int) Math.ceil(originalFr.numRows() / 1e3), H2O.NUMCPUS);
   }
 
+  // multi-node version (experimental version)
+  private int desiredChunk_multi(final Frame fr) {
+    for (int type : fr.types()) {
+      if (type != Vec.T_NUM && type != Vec.T_CAT) {
+        Log.warn("Training frame contains columns non-numeric/categorical columns. Using old rebalance logic.");
```

Comment:
Why the new logic cannot handle String columns?

Reply:
It could, there is no reason why not. It adds complexity and requires more testing which I wanted to avoid in the PR (it should go to a fix release).

The rebalance is done after the String/UUID/.. columns are already removed for most algos (expect for word2vec). For most algos there is no issue. Worst case is it will behave the old way for word2vec.

I will add the support to the master version of this branch in PUBDEV-5538. WDYT?